pkg/languagedetection/util/languageset.go

```diff
@@ -8,80 +8,98 @@ package util
 type LanguageSet map[Language]struct{}
 
-// DeepCopy returns a deep copy of the current language set
-func (s LanguageSet) DeepCopy() LanguageSet {
-	copyLanguageSet := LanguageSet{}
+// Add adds a new language to the language set
+// returns false if the language is already included in the set, and true otherwise
+func (s LanguageSet) Add(language Language) bool {
+	_, found := s[language]
+	s[language] = struct{}{}
+	return !found
+}
 
+// ToProto returns a proto message Language
+func (s LanguageSet) ToProto() []*pbgo.Language {
+	res := make([]*pbgo.Language, 0, len(s))
 	for lang := range s {
-		copyLanguageSet[lang] = struct{}{}
+		res = append(res, &pbgo.Language{
+			Name: string(lang),
+		})
 	}
+	return res
+}
+
+////////////////////////////////
+//                            //
+//     Timed Language Set     //
+//                            //
+////////////////////////////////
 
-	return copyLanguageSet
+// TimedLanguageSet handles storing sets of languages along with their expiration times
+type TimedLanguageSet map[Language]time.Time
+
+// RemoveExpired removes all expired languages from the set
+// Returns true if at least one language is expired and removed
+func (s TimedLanguageSet) RemoveExpired() bool {
+	removedAtLeastOne := false
+	for lang, expiration := range s {
+		if expiration.Before(time.Now()) {
+			s.Remove(lang)
+			removedAtLeastOne = true
+		}
+	}
+	return removedAtLeastOne
 }
 
-// Add adds a new language to the language set
-// returns false if the language is already
-// included in the set, and true otherwise
-func (s LanguageSet) Add(language Language) bool {
+// Add adds a new language to the language set with an expiration time
+// returns false if the language is already included in the set, and true otherwise
+func (s TimedLanguageSet) Add(language Language, expiration time.Time) bool {
 	_, found := s[language]
-	s[language] = struct{}{}
+	s[language] = expiration
 	return !found
 }
 
 // Has returns whether the set contains a specific language
-func (s LanguageSet) Has(language Language) bool {
+func (s TimedLanguageSet) Has(language Language) bool {
 	_, found := s[language]
 	return found
 }
 
 // Remove deletes a language from the language set
-func (s LanguageSet) Remove(language Language) {
+func (s TimedLanguageSet) Remove(language Language) {
 	delete(s, language)
 }
 
-// Merge merges another language set with the current language set
-// returns true if the set has changed after merge, and false otherwise
-func (s LanguageSet) Merge(other LanguageSet) bool {
-	if len(other) == 0 {
-		return false
-	}
+// Merge merges another timed language set with the current language set
+// returns true if the set new languages were introduced, and false otherwise
+func (s TimedLanguageSet) Merge(other TimedLanguageSet) bool {
 
-	sizeBeforeMerge := len(s)
-	for language := range other {
-		s.Add(language)
+	modified := false
+
+	for language, expiration := range other {
+		if s.Add(language, expiration) {
```

Comment:
Yes this is intended and I made sure it is clearly stated in the method comment:

`// returns false if the language is already included in the set, and true otherwise`

We don't consider that the language set is changed just because we refreshed a TTL.

This is mainly because TTLs are not stored in workloadmeta, so we don't care if they changed or not, we only care about languages existing in the set